include/rocksdb/advanced_iterator.h

```diff
@@ -0,0 +1,34 @@
+//  Copyright (c) Meta Platforms, Inc. and affiliates.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include "rocksdb/slice.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+enum class IterBoundCheck : char {
+  kUnknown = 0,
+  kOutOfBound,
+  kInbound,
+};
+
+// This structure encapsulates the result of NextAndGetResult()
+struct IterateResult {
+  Slice key;
```

Comment:
It seems like we have an opportunity to do something like IOBuf in the handling of keys and values. In many/most cases we will have materialized a key, and maybe in some cases a value, that we should be able to hand off to the user to make use of as long as they want without deep copying. (Perhaps "unnecessary copying" is more of a CPU overhead for scans than "unnecessary virtual calls". Hmm.)

Probably not a simple change, but perhaps something for the TODO list.